neuralop/data/datasets/web_utils.py

```diff
@@ -0,0 +1,160 @@
+import sys
+from typing import Any, Optional, Union, List
+import tarfile
+
+logger = logging.Logger(name=__name__)
+logger.setLevel(logging.root.level)
+
+# md5 utils from torchvision.datasets.utils
+
+def calculate_md5(fpath: Union[str, Path], chunk_size: int = 1024 * 1024) -> str:
+    # Setting the `usedforsecurity` flag does not change anything about the functionality, but indicates that we are
+    # not using the MD5 checksum for cryptography. This enables its usage in restricted environments like FIPS. Without
+    # it torchvision.datasets is unusable in these environments since we perform a MD5 check everywhere.
+    if sys.version_info >= (3, 9):
+        md5 = hashlib.md5(usedforsecurity=False)
+    else:
+        md5 = hashlib.md5()
+    with open(fpath, "rb") as f:
+        while chunk := f.read(chunk_size):
+            md5.update(chunk)
+    return md5.hexdigest()
+
+def check_md5(fpath: Union[str, Path], md5: str, **kwargs: Any) -> bool:
+    return md5 == calculate_md5(fpath, **kwargs)
+
+def check_integrity(fpath: Union[str, Path], md5: Optional[str] = None) -> bool:
+    if isinstance(fpath, str):
+        fpath = Path(fpath)
+    if not fpath.is_file():
+        return False
+    if md5 is None:
+        return True
+    return check_md5(fpath, md5)
+
+def download_from_url(
+    url: str,
+    root: Union[str, Path],
+    filename: Optional[Union[str, Path]] = None,
+    md5: Optional[str] = None,
+    size: Optional[int] = None,
+    chunk_size: Optional[int] = 256 * 64,
+    extract_tars: bool = True
+) -> None:
+    """download_from_url downloads a file from a url with 
+    an optional md5 checksum. 
+    
+    If the file is a tarball, optionally extract.
+
+    Parameters
+    ----------
+    url : str
+        url where dataset archive file is stored
+    root : Union[str, Path]
+        root of dataset folder on local machine
+    filename : Optional[Union[str, Path]], optional
+        name into which to download, by default None
+    md5 : Optional[str], optional
+        md5 checksum to verify file correctness, by default None
+    size: Optional[int], optional
+        size of the file expected in bytes, to check
+    chunk_size : int, optional
+        size of chunks to use when streaming files
+    extract_tars: bool, optional
+        whether to extract .tgz archives, by default True
+    """
+    if isinstance(root, str):
+        root = Path(str)
+    
+    root = root.expanduser()
+    if not filename:
+        # grab file ext from basename
+        filename = url.split('/')[-1] 
+    fpath = root / filename
+
+    root.mkdir(parents=True, exist_ok=True)
+
+    # check if file is already present locally
+    if check_integrity(fpath, md5):
+        logger.info("Using downloaded and verified file: " + str(fpath))
+        return
+
+    # download and stream file in chunks
+    with requests.get(url, stream=True) as r:
+        r.raise_for_status()
+        confirm = input("Confirm download [Y/n]: ")
```

Comment:
One minor comment: I'm not sure we want/should ask for input. This is helpful locally but at scale or on clusters may quickly become a hinderance. 

I'd just make sure we abide by sensitive flags in downstream functions (e.g. `download=True`)

Reply:
Thanks! Removed them.